server/src/main/java/org/elasticsearch/script/ScriptEngine.java

```diff
@@ -45,4 +45,15 @@
 
     @Override
     default void close() throws IOException {}
+
+    /**
+     * Usually scripts require compilation. However there can be special cases where scripts do not require any compilation to happen.
+     * For example, when the input is equal to the output when rendering a string based template without parameters.
+     *
+     * @param scriptSource the script
+     * @return             true if it makes sense to put the executed script in the cache, false otherwise
+     */
+    default boolean requiresCompilation(String scriptSource) {
```

Comment:
I think this terminology and workflow is confusing, since the name would imply compile won't be called, but that is exactly what we still do.

What if instead we were to encapsulate the cache and move to an argument to compile, which each script engine may choose to use? Inserting into the cache would then increment the counter and trigger the limit check.

Reply:
> What if instead we were to encapsulate the cache and move to an argument to compile, which each script engine may choose to use? Inserting into the cache would then increment the counter and trigger the limit check.

👍

Reply:
I played around a little bit today and had a quick chat with Martijn, but I failed to see a good solution. Here is the culprit: We actually have to take care of two checks, one before the compilation calling `checkCompilationLimit` and putting things into the cache after compilation. While it would be easy to add two consumers/functions to the compile method, I dont think this is a good approach, as I really dont want to put it on every single script engine implementation to call `checkCompilationLimit()`.

I may have missed something, so eager to find the solution I missed out.